### PR TITLE
Multi profile

### PR DESCRIPTION
**Pull request by e-davidson**
add ability to have addition profiles to save on top of the default 
change component store to use mixin

### CODE REVIEW

File: projects/table-builder/src/lib/functions/rxjs-operators.ts [hunk 1]
```diff
@@ -31,3 +31,45 @@ export const combineArrays = <T>(sources: Observable<T[]>[]): Observable<T[]> =>
     map( res => _.flatten(res) )
   );
 };
+
+export function switchOff( switchSource: Observable<boolean>, defaultState: boolean = true) {
+  return <T>(source: Observable<T>) : Observable<T> => {
+    return new Observable(subsciber => {
+      let isOn = defaultState;
+      const subscription = new Subscription();
+      subscription.add( switchSource.subscribe( on => isOn = on ));
```
Comment (NachmanBerkowitz) [inline projects/table-builder/src/lib/functions/rxjs-operators.ts:40]: I would assume that there is no guarantee as to the timing. The `next()` on the `switchSource` maybe be delayed (with a `scheduler` or something

File: projects/table-builder/src/lib/components/table-container/table-container.html [hunk 2]
```diff
@@ -7,17 +7,54 @@
...
+            <span>New</span>
+          </button>
+          <ng-container *ngIf="currentStateKey$ | async as currentKey" >
+            <ng-container  *ngFor='let key of stateKeys$ | async'>
+              <button mat-menu-item (click)='setProfileState(key)' >
+                <div style='display: flex; align-items: center; justify-content: space-between;'>
+                  <span style='display:flex;'>{{key}}</span>
+                  <span style='display:flex;'>
+                    <span style="width: 120px;"></span>
+                    <mat-icon color='warn' (click)='deleteProfileState(key)' stop-propagation >delete_forever</mat-icon>
```
Comment (NachmanBerkowitz) [inline projects/table-builder/src/lib/components/table-container/table-container.html:41]: Is it possible not to shaow delete icon for pending?

File: projects/table-builder/src/lib/classes/component-store-extensions.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+interface ConcreteClass<C> { new (...args: any[]): C; }
+
+
+interface ComponentStoreType<T extends {}> extends ConcreteClass<ComponentStore<T>> {
+
+}
+
+export function ComponentStoreExtensions<T extends {}>(){
+  return function <U extends ComponentStoreType<T>>(constructor: U) {
```
[the remark this answers is not in the document]
Comment (e-davidson) [reply projects/table-builder/src/lib/classes/component-store-extensions.ts:14]: Short answer: Mixins give you more flexibility than inheritance.

[the remark this answers is not in the document]
Comment (e-davidson) [reply projects/table-builder/src/lib/classes/component-store-extensions.ts:14]: This is a typescript limitation for doing mixins with generics.
I plan on filing an issue.
See this stack overflow
https://stackoverflow.com/questions/48492002/passing-generic-types-through-to-mixins-in-typescript

File: projects/table-builder/src/lib/classes/table-store.ts [hunk 4]
```diff
@@ -5,50 +5,40 @@ import { defaultTableState, TableState } from './TableState';
...
-  getFilter$ = (filterId) : Observable<FilterInfo | undefined> => {
-    return this.select( this.filters$, filters => filters[filterId]);
+  constructor(@Inject(TableBuilderConfigToken) config: TableBuilderConfig) {
+   super( { ...defaultTableState, ...config.defaultTableState});
   }
 
+  getSavableState() : Observable<TableState> {
+    return  this.state$.pipe(
+      map( s => {
+        const metaData = Object.values(s.metaData)
```
[the remark this answers is not in the document]
Comment (e-davidson) [reply projects/table-builder/src/lib/classes/table-store.ts:30]: good catch.
I was debating whether there was ever any point in saving it at all.

File: projects/table-builder/src/lib/functions/rxjs-operators.ts (hunk 1 above)
Comment (e-davidson) [reply projects/table-builder/src/lib/functions/rxjs-operators.ts:40]: good point, but I'm not really sure what else can be done.
Like is there a way to force them both onto the same scheduler? 
Any suggestions?

File: projects/table-builder/src/lib/components/table-container/table-container.html [hunk 5]
```diff
@@ -7,17 +7,54 @@
...
-        </button>
-        <button *ngIf="tableId" mat-icon-button (click)="state.saveToState(tableId);OnSaveState.next()" matTooltip="Save Table">
-          <mat-icon color="primary">save</mat-icon>
-        </button>
-        <button mat-icon-button (click)="exportToCsv()" matTooltip="Export Table">
-          <mat-icon color="primary">file_download</mat-icon>
-        </button>
+        <button mat-icon-button color='primary' [matMenuTriggerFor]="mainMenu"><mat-icon>more_vert</mat-icon></button>
+        <mat-menu #mainMenu='matMenu' >
+          <button mat-menu-item (click)="state.resetState();OnStateReset.next()" >
```
[the remark this answers is not in the document]
Comment (e-davidson) [reply projects/table-builder/src/lib/components/table-container/table-container.html:12]: possible yes, but not for now.

File: projects/table-builder/src/lib/components/table-container/table-container.html (hunk 2 above)
Comment (e-davidson) [reply projects/table-builder/src/lib/components/table-container/table-container.html:41]: what does pending mean to  you? and why ?

Comment (NachmanBerkowitz) [reply projects/table-builder/src/lib/components/table-container/table-container.html:41]: I meant for Default

File: projects/table-builder/src/lib/classes/table-store.ts [hunk 6]
```diff
@@ -57,7 +47,7 @@ export class TableStore extends ComponentStore<TableState> {
   );
 
   getMetaData$ = (key: string) : Observable<MetaData> => {
-    return this.select(this.metaData$, md => md.find(m => m.key === key ))
+    return this.select( state => state.metaData[key]  )
   }
 
   createPreSort = (metaDatas: Dictionary<MetaData>): Sort[] => {
```
Comment (NachmanBerkowitz) [inline projects/table-builder/src/lib/classes/table-store.ts:53]: `mergeMetaDatas` methos seems to be incorrect. It should be
` mergeMetaDatas = (existingMetaData: Dictionary<MetaData>, incomingMetaData: Dictionary<MetaData>) => {
    const metaData: Dictionary<MetaData> = {};
    const incomingMetaDatas = Object.values(incomingMetaData);
    incomingMetaDatas.forEach( md => {
        const existing = metaData[md.key] ?? existingMetaData[md.key];
        if(!existing) {
          metaData[md.key] = { ...md, noExport: md.customCell }
        } else {
          metaData[md.key] = this.mergeMeta(existing,md);
        }
    });
    return metaData;
  }`

Comment (e-davidson) [reply projects/table-builder/src/lib/classes/table-store.ts:54]: I think this should fix it, although I suspect there may be a bug as now it doesn't auto add the custom cell field

File: projects/table-builder/src/lib/classes/table-store.ts [hunk 7]
```diff
@@ -128,18 +119,19 @@ export class TableStore extends ComponentStore<TableState> {
...
-        if(!existing) {
-          metaData[md.key] = { ...md, noExport: md.customCell }
-        } else {
-          metaData[md.key] = this.mergeMeta(existing,md);
-        }
-    });
-    return {...metaData, ...newMetaDatas};
+  mergeMetaDatas = (existingMetaData: Dictionary<MetaData>, incomingMetaData: Dictionary<MetaData>) : Dictionary<MetaData> => {
+    const allKeys = Object.keys(existingMetaData).concat(Object.keys(incomingMetaData));
+    const keys: string [] = _.uniq(allKeys, false);
```
Comment (NachmanBerkowitz) [inline projects/table-builder/src/lib/classes/table-store.ts:124]: Maybe `keys = [...new Set(Object.keys(existingMetaData).concat(Object.keys(incomingMetaData))]`

**Review by NachmanBerkowitz — APPROVED** [review]

**Pull request merged**